packages/base-controller/src/ControllerMessenger.ts

```diff
@@ -1,32 +1,48 @@
-export type ActionHandler<Action, ActionType> = (
+export type ActionHandler<
+  Action extends ActionConstraint,
+  ActionType = Action['type'],
+> = (
   ...args: ExtractActionParameters<Action, ActionType>
 ) => ExtractActionResponse<Action, ActionType>;
-export type ExtractActionParameters<Action, T> = Action extends {
-  type: T;
-  handler: (...args: infer H) => any;
+
+export type ExtractActionParameters<
+  Action extends ActionConstraint,
+  ActionType = Action['type'],
+> = Action extends {
+  type: ActionType;
+  handler: (...args: infer HandlerArgs) => unknown;
 }
-  ? H
+  ? HandlerArgs
```

Comment:
Continuing this thread: https://github.com/MetaMask/core/pull/1890#discussion_r1373773411